lib/transaction.js

```diff
@@ -200,6 +154,64 @@ class Transaction extends EventEmitter {
     return this;
   }
 
+  async _kickThingsOff(config, container) {
+    // FYI: This is temporarily stalling things so that the constructor
```

Comment:
maybe move _previousSibling higher than this._kickThingsOff?

Reply:
It's a bit of a chicken-and-egg problem because `_previousSibling` is derived from `basePromise` (Well, more precisely: the previous Transaction's `basePromise`).  My concern is that this will introduce a subtle race condition that can cause `outerTx._lastChild` to point to the wrong `Promise`.

But... we might be able to address that by changing how/when `outerTx._lastChild` gets assigned.  I'll have to think about that a bit more.

Reply:
Saying this another way: I think this relates to a larger problem that I'm seeing throughout the `Transaction` class.

There are currently several places throughout the `Transaction` class where the `Transaction` needs to be aware of how/when it is being constructed/used.  In this case, the `Transaction` needs to change its behavior depending upon whether or not there was an `outerTx`.

So, one of my main goals in refactoring the `Transaction` class is to move these responsibilities into a higher layer of the code.  But, there will probably be 2-3 separate PRs before this goal is completed.

Reply:
Here's what I'm planning for some upcoming PRs:

As mentioned previously, the `Transaction` class often needs to change its behavior depending upon how/when it is being constructed/used.  However, it does not have any direct way to know this information.  Instead, it infers the details by:

1. Checking for the presence of `outerTx`.  (Let's the `Transaction` know whether or not it is a "top-level" transaction)
2. Checking for the presence of `config.connection`.  (This seems to be another way for the `Transaction` to determine whether it is top-level or not)
3. Checking to see if `container` returned a `Promise`.  (Helps the `Transaction` decide if it's responsible for the commit/rollback operations)

It looks like most of these details can be extracted into the Factory components that are instantiating the `Transactions`.  More specifically: the `Factory` components can pass in the appropriate strategy functions / polymorphic objects when instantiating the `Transaction`s.

For example: right now, there are several places in the code where `Transaction` changes its behavior depending upon whether or not `outerTx` was defined.  Instead, the code will be revised so that `outerTx` is _always_ defined as one of two implementations:

1. One implementation that is used by "Top-Level" transactions
2. A second implementation that is used by "Nested" transactions

So, how do we decide which implementation to use?  From what I can tell, it's actually fairly straightforward:

1. Factories instantiated through `makeKnex` always use the "Top-Level" implementation
2. Factories instantiated through `makeTransactor` always use the "Nested" implementation

(Note: right now, the main difference between these two Factories seems to be whether or not `outerTx` is passed into the `Transaction` constructor.  So, this seems to align nicely w/ this proposal)

Once this change is in place, it becomes possible for the `Transaction` class to delegate various decisions to this object.  For example:

1. Does the `connection` need to be disposed once the `Transaction` has concluded?
2. Does the `Transaction` need to register itself as a sibling?
3. Etc.

That's the high-level idea, at least.  But, as you probably guessed, this also means that this is going to change the API for the `Transaction` constructor.  This is a somewhat-breaking change, but fortunately it appears that the `Transaction` constructor is never actually called directly.

Anyhoo: this is my main reason for leaving the "Temporarily stall things" hack in place for now.  We can go ahead and fix it in this PR, but it also means that the scope of this PR will increase quite a bit.  I'd rather merge this minor enhancement/fix first, and then tackle the larger code restructuring afterwards.

cc: @maximelkin + @kibertoad

Reply:
@briandamaged nice!

Reply:
So, an interesting discovery while investigating some of these ideas: it appears that `config.connection` is only used by `sqlite3`.  From what I can tell, it's being used for 2 reasons:

1. To facilitate a handful of `sqlite3` unit tests. (ie: allow the unit tests to bypass certain logic)
2. To ensure that there is only ever 1 connection created.  (it looks like `sqlite3` might not work properly when there is more than one connection in the pool)

As an experiment, I removed the `config.connection` logic.  Most things seem to work fine, but a handful of `sqlite3` operations break because they are unable to obtain a connection (because the pool max is 1).  So, I'm doing a bit more investigation to figure out why `sqlite3` is attempting to create new connections in those specific cases.